packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -2754,7 +2754,10 @@ abstract class BuildableElement extends Element {
     if (dirty)
       return;
     _dirty = true;
-    owner.scheduleBuildFor(this);
+    if (_inDirtyList)
+      owner.markNeedsToResortDirtyElements();
+    else
+      owner.scheduleBuildFor(this);
```

Comment:
We can simplify the code further by moving this logic into scheduleBuildFor. It would end up removing the assert entirely, I think.

What we're losing here is any sanity-checking that we're not marking things dirty when they've already been cleaned. The previous approach was to special-case each time we ran into a case that could trigger that.